cf/data/data.py

```diff
@@ -6366,42 +6366,22 @@ def mask(self):
         (12, 73, 96)
         >>> m = d.mask
         >>> m.dtype
-        dtype('bool')
+        bool
         >>> m.shape
-        (12, 73, 96])
+        (12, 73, 96)
 
         """
-        mask = self.copy()
-
-        config = mask.partition_configuration(readonly=False)
-
-        for partition in mask.partitions.matrix.flat:
-            partition.open(config)
-            array = partition.array
-
-            if partition.masked:
-                # Array is masked
-                partition.subarray = array.mask.copy()
-            else:
-                # Array is not masked
-                partition.subarray = FilledArray(
-                    shape=array.shape,
-                    size=array.size,
-                    ndim=array.ndim,
-                    dtype=_dtype_bool,
-                    fill_value=0,
-                )
-
-            partition.Units = _units_None
-
-            partition.close()
+        mask_data_obj = self.copy()
 
-        mask._Units = _units_None
-        mask.dtype = _dtype_bool
+        dx = self._get_dask().copy()
+        mask = da.ma.getmaskarray(dx)
+        mask_data_obj[...] = mask
 
-        mask._hardmask = True
+        mask_data_obj.dtype = _dtype_bool
+        mask_data_obj._Units = _units_None
+        mask_data_obj._hardmask = True
 
-        return mask
+        return mask_data_obj
```

Comment:
This might not work in all cases, e.g. I'm not sure what happens is `self` is an array of strings. How about:
```python
mask_data_obj = self.copy()

dx = self._get_dask()
mask = da.ma.getmaskarray(dx)
mask_data_obj._set_dask(mask, reset_mask_hardness=True)
mask_data_obj.override_units(_units_None, inplace=True)
mask_data_obj.hardmask = True

return mask_data_obj
```
This is more efficient, too, as there is no `__setitem__` nor `astype` operation. We need to set `hardmask` (without the underscore) so that the `cf_harden_mask` operation is added to the dask graph. The data type is automatically taken from the underlying new Boolean dask array.

Reply:
Aha, much better idea than `[...]`, thanks!

I did first have a poke through the code and docs to check we didn't have something that could effectively replace all of the data with new data, rather than setting it all to something new (if the difference makes sense?), so it is strange I didn't arrive at `_set_dask` from that!

Reply:
> I'm not sure what happens is self is an array of strings.

Sounds like we should add a dedicated test and cover this. I was thinking of extending `test_Data_apply_masking` to first test `.mask`.